src/control/lib/control/pool.go

```diff
@@ -1057,89 +1063,212 @@ func processSCMSpaceStats(log debugLogger, filterRank filterRankFn, scmNamespace
 			scmBytes = usableBytes
 		}
 
-		if _, exists := rankNVMeFreeSpace[scmNamespace.Mount.Rank]; exists {
-			return 0, errors.Errorf("Multiple SCM devices found for rank %d",
+		if _, exists := req.rankNVMeFreeSpace[scmNamespace.Mount.Rank]; exists {
+			return 0, errors.Errorf("multiple scm devices found for rank %d",
 				scmNamespace.Mount.Rank)
 		}
 
 		// Initialize entry for rank in NVMe free space map.
-		rankNVMeFreeSpace[scmNamespace.Mount.Rank] = 0
+		req.rankNVMeFreeSpace[scmNamespace.Mount.Rank] = 0
 	}
 
 	return scmBytes, nil
 }
 
+// Check whether controller on a storage server shares both META and DATA roles. If so then META
+// capacity needs to be subtracted from available capacity calculated to store DATA.
+//
+// Assumptions made during this check:
+// - Dual engines on host contain identical bdev tier configurations in storage config
+// - Each role (wal/meta/data) can only be assigned to one bdev tier
+//
+// Extrapolating from these assumptions it can be asserted that if ONE of a host's controllers
+// has been assigned both META and DATA roles then all other host SSDs with DATA role assigned
+// will also hold metadata.
+func areMetaDataRolesShared(nvmeControllers storage.NvmeControllers) bool {
+	for _, controller := range nvmeControllers {
+		for _, smdDevice := range controller.SmdDevices {
+			if smdDevice.Roles.HasMeta() && smdDevice.Roles.HasData() {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
 // Add NVMe free bytes to rankNVMeFreeSpace map.
-func processNVMeSpaceStats(log debugLogger, filterRank filterRankFn, nvmeControllers storage.NvmeControllers, rankNVMeFreeSpace rankFreeSpaceMap) error {
+func processNVMeSpaceStats(req processSpaceReq, nvmeControllers storage.NvmeControllers) error {
 	for _, controller := range nvmeControllers {
 		for _, smdDevice := range controller.SmdDevices {
-			msgDev := fmt.Sprintf("SMD device %s (rank %d, ctrlr %s)", smdDevice.UUID,
+			msgDev := fmt.Sprintf("smd device %s (rank %d, ctrlr %s", smdDevice.UUID,
 				smdDevice.Rank, controller.PciAddr)
 
-			if !smdDevice.Roles.IsEmpty() && (smdDevice.Roles.OptionBits&storage.BdevRoleData) == 0 {
-				log.Debugf("Skipping %s, not used for storing data", msgDev)
-				continue
+			if smdDevice.Roles.IsEmpty() {
+				msgDev += ")"
```

Comment:
Wouldn't we still want to skip in this case?

Reply:
no, if empty it just means we are in PMem mode

Reply:
Ah, understood.